### ISSUE TITLE
Mounting Errors.

### ISSUE THREAD

**Issue by wojtek-viirtue** [kadalu/kadalu#468]
**Getting the following errors on various kadalu-csi-nodeplugin pods:**

```
Loading...
[2021-03-06 16:18:47,959] INFO [kadalulib - 334:start_all] - Started Process	 name=csi
[2021-03-06 16:18:47,961] INFO [kadalulib - 334:start_all] - Started Process	 name=metrics
[2021-03-06 16:18:48,888] INFO [exporter - 85:<module>] - Started Kadalu Storage CSI Metrics exporter.	 port=8000
[2021-03-06 16:18:49,095] INFO [main - 34:main] - Server started
[2021-03-06 16:19:51,979] INFO [nodeserver - 83:NodePublishVolume] - Mounted PV	 volume=pvc-bf964181-a99e-4cac-8e57-bbeb1e2a74e9 pvpath=subvol/a2/84/pvc-bf964181-a99e-4cac-8e57-bbeb1e2a74e9 pvtype=subvol hostvol=storage-pool-1 duration_seconds=0.6030356884002686
[2021-03-06 16:29:12,868] INFO [nodeserver - 83:NodePublishVolume] - Mounted PV	 volume=pvc-bfa06dd8-8d2b-4a83-91c3-1cfe2fcc2cff pvpath=subvol/fe/3a/pvc-bfa06dd8-8d2b-4a83-91c3-1cfe2fcc2cff pvtype=subvol hostvol=storage-pool-1 duration_seconds=0.026730060577392578
[2021-03-06 16:29:21,215] INFO [nodeserver - 83:NodePublishVolume] - Mounted PV	 volume=pvc-a5c1bede-fe0e-4378-b97d-c49e0d8ffce7 pvpath=subvol/5a/5b/pvc-a5c1bede-fe0e-4378-b97d-c49e0d8ffce7 pvtype=subvol hostvol=storage-pool-1 duration_seconds=0.027896404266357422
[2021-03-06 17:58:22,898] ERROR [volumeutils - 862:mount_glusterfs] - error to execute command	 volume={'name': 'storage-pool-1', 'g_volname': None, 'g_host': None, 'g_options': None, 'type': 'Replica3'} cmd=['/opt/sbin/glusterfs', '--process-name', 'fuse', '-l', '/var/log/gluster/gluster.log', '--volfile-id', 'storage-pool-1', '--fs-display-name', 'kadalu:storage-pool-1', '-f', '/kadalu/volfiles/storage-pool-1.client.vol', '/mnt/storage-pool-1'] error=[1]
[2021-03-06 17:58:22,898] ERROR [_server - 445:_call_behavior] - Exception calling application: [1]
Traceback (most recent call last):
File "/kadalu/lib/python3.8/site-packages/grpc/_server.py", line 435, in _call_behavior
response_or_iterator = behavior(argument, context)
File "/kadalu/nodeserver.py", line 72, in NodePublishVolume
mount_glusterfs(volume, mntdir, True)
File "/kadalu/volumeutils.py", line 868, in mount_glusterfs
raise err
File "/kadalu/volumeutils.py", line 860, in mount_glusterfs
execute(*cmd)
File "/kadalu/kadalulib.py", line 121, in execute
raise CommandException(proc.returncode, out.strip(), err.strip())
kadalulib.CommandException: [1]
```

**The error is correlated to PVCs not being able to mount:**

```
Events:
  Type     Reason       Age                   From               Message
  ----     ------       ----                  ----               -------
  Normal   Scheduled    6m37s                 default-scheduler  Successfully assigned gocd/connect-to-gocd-pv-pod to viirtue-dev-node-worker-1
  Warning  FailedMount  24s (x11 over 6m37s)  kubelet            MountVolume.SetUp failed for volume "pvc-bf964181-a99e-4cac-8e57-bbeb1e2a74e9" : rpc error: code = Unknown desc = Exception calling application: [1]
  Warning  FailedMount  6s (x3 over 4m34s)    kubelet            Unable to attach or mount volumes: unmounted volumes=[task-pv-storage], unattached volumes=[task-pv-storage default-token-gzfmx]: timed out waiting for the condition
```

If I restart the offending kadalu-csi-nodeplugin pod and then the pod that was trying to mount....it works. It'll be good for almost an hour and then fail again

Here is an output of the healinfo which doesn't show any abnormalities:

```
Giving heal information of volume storage-pool-1
Brick server-storage-pool-1-0-0.storage-pool-1:/bricks/storage-pool-1/data/brick
Status: Connected
Total Number of entries: 0
Number of entries in heal pending: 0
Number of entries in split-brain: 0
Number of entries possibly healing: 0

Brick server-storage-pool-1-1-0.storage-pool-1:/bricks/storage-pool-1/data/brick
Status: Connected
Total Number of entries: 0
Number of entries in heal pending: 0
Number of entries in split-brain: 0
Number of entries possibly healing: 0

Brick server-storage-pool-1-2-0.storage-pool-1:/bricks/storage-pool-1/data/brick
Status: Connected
Total Number of entries: 0
Number of entries in heal pending: 0
Number of entries in split-brain: 0
Number of entries possibly healing: 0
```

Any help would be greatly appreciated

k8s v1.19.4, 
kadalu: 0.8.0

**Comment by amarts** [comment]
Is there any logs in csi-nodeplugin for glusterfs? Can you do get logs of all-containers in csi-nodeplugin pod? That would help us to understand the reason for 'failure' of mount!

**Comment by wojtek-viirtue** [comment]
It just happend again.....when the error condition happens, the container using the PVC never terminates and I'm forced to kill it using the following: 
"kubectl delete pod gocd-server-7d4f8dfc8c-gswqb --grace-period=0 --force". (I verify that it's actually not running still in containerd.

I'm afraid that the errors I'm showing you are the product of what happens after the issue occurs (stale mount)...but at some point...it looks like the container is losing access to the mount and then when it's restarted...the errors are produced.

When trying to restart the deployment/pod that uses the pvc the csi-node-plugin just outputs the following from the **logging container:**

```
[2021-03-06 18:10:41.592864 +0000] I [MSGID: 114057] [client-handshake.c:1126:select_server_supported_programs] 0-storage-pool-1-client-1: Using Program [{Program-name=GlusterFS 4.x v1}, {Num=1298437}, {Version=400}] 
[2021-03-06 18:10:41.592899 +0000] I [MSGID: 114057] [client-handshake.c:1126:select_server_supported_programs] 0-storage-pool-1-client-0: Using Program [{Program-name=GlusterFS 4.x v1}, {Num=1298437}, {Version=400}] 
[2021-03-06 18:10:41.592968 +0000] I [MSGID: 114057] [client-handshake.c:1126:select_server_supported_programs] 0-storage-pool-1-client-2: Using Program [{Program-name=GlusterFS 4.x v1}, {Num=1298437}, {Version=400}] 
[2021-03-06 18:10:41.593545 +0000] I [MSGID: 114046] [client-handshake.c:855:client_setvolume_cbk] 0-storage-pool-1-client-0: Connected, attached to remote volume [{conn-name=storage-pool-1-client-0}, {remote_subvol=/bricks/storage-pool-1/data/brick}] 
[2021-03-06 18:10:41.593565 +0000] I [MSGID: 108005] [afr-common.c:6053:__afr_handle_child_up_event] 0-storage-pool-1-replica-0: Subvolume 'storage-pool-1-client-0' came back up; going online. 
[2021-03-06 18:10:41.593754 +0000] I [MSGID: 114046] [client-handshake.c:855:client_setvolume_cbk] 0-storage-pool-1-client-1: Connected, attached to remote volume [{conn-name=storage-pool-1-client-1}, {remote_subvol=/bricks/storage-pool-1/data/brick}] 
[2021-03-06 18:10:41.593918 +0000] I [MSGID: 114046] [client-handshake.c:855:client_setvolume_cbk] 0-storage-pool-1-client-2: Connected, attached to remote volume [{conn-name=storage-pool-1-client-2}, {remote_subvol=/bricks/storage-pool-1/data/brick}] 
[2021-03-06 18:10:41.594026 +0000] I [MSGID: 108002] [afr-common.c:6425:afr_notify] 0-storage-pool-1-replica-0: Client-quorum is met 
[2021-03-06 18:10:41.595554 +0000] I [fuse-bridge.c:5315:fuse_init] 0-glusterfs-fuse: FUSE inited with protocol versions: glusterfs 7.24 kernel 7.31
[2021-03-06 18:10:41.595571 +0000] I [fuse-bridge.c:5947:fuse_graph_sync] 0-fuse: switched to graph 0
[2021-03-06 18:10:42.617727 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 18:30:19.118788 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 18:30:19.325503 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 18:35:19.853593 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 18:40:19.992333 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 18:45:20.156809 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 18:50:20.323108 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 18:55:20.478139 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:00:20.632001 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:05:20.774446 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:10:20.911057 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:15:21.046262 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:20:21.180316 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:25:21.260818 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:30:21.260828 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:35:21.260827 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:40:21.260816 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:45:21.260861 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:50:21.260829 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 19:55:21.260816 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:00:21.260816 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:05:21.260853 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:10:21.260816 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:15:21.260819 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:20:21.260836 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:25:21.260819 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:30:21.204296 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:35:21.260840 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:40:21.260814 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:45:21.260808 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:50:21.260806 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 20:55:21.260806 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:00:21.260824 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:05:21.260815 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:10:21.260817 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:15:21.260828 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:20:21.260797 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:25:21.260818 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:30:21.260823 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:35:21.260840 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:40:21.260823 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:45:21.260809 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:50:21.260828 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 21:55:21.260827 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:00:21.260806 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:05:21.038003 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:10:21.260830 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:15:21.260831 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:20:21.260808 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:25:21.260831 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:30:21.260859 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:35:21.104169 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:40:21.244885 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:45:21.260843 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:50:21.260809 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 22:55:21.260804 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:00:21.260825 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:05:21.260828 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:10:21.260818 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:15:21.260819 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:20:21.260836 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:25:21.260810 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:30:21.260836 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:35:21.260821 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:40:21.083815 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:45:21.260793 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:50:21.260811 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-06 23:55:21.260832 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-07 00:00:21.260791 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-07 00:05:21.260823 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-07 00:10:21.261793 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-07 00:15:21.260873 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-07 00:20:21.260830 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-07 00:25:21.260797 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-07 00:30:21.260855 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-03-07 00:35:21.260827 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
pending frames:
frame : type(0) op(0)
frame : type(0) op(0)
patchset: git://git.gluster.org/glusterfs.git
signal received: 11
time of crash: 
2021-03-07 00:38:30 +0000
configuration details:
argp 1
backtrace 1
dlfcn 1
libpthread 1
llistxattr 1
setfsid 1
spinlock 1
epoll.h 1
xattr.h 1
st_atim.tv_nsec 1
package-string: glusterfs 2021.03.02
/opt/lib/libglusterfs.so.0(+0x2bc54)[0x7f87729d1c54]
/opt/lib/libglusterfs.so.0(gf_print_trace+0x486)[0x7f87729dd0b6]
/lib/x86_64-linux-gnu/libc.so.6(+0x46210)[0x7f877277a210]
/opt/lib/libglusterfs.so.0(__gf_free+0xb0)[0x7f87729fb440]
/opt/lib/glusterfs/2021.03.02/xlator/cluster/replicate.so(+0x67f40)[0x7f876e4edf40]
/opt/lib/glusterfs/2021.03.02/xlator/cluster/replicate.so(+0x67fc8)[0x7f876e4edfc8]
/opt/lib/glusterfs/2021.03.02/xlator/cluster/replicate.so(+0x67fdd)[0x7f876e4edfdd]
/opt/lib/libglusterfs.so.0(fd_unref+0x11f)[0x7f87729f8c1f]
/opt/lib/glusterfs/2021.03.02/xlator/mount/fuse.so(+0x203d3)[0x7f87715e53d3]
/opt/lib/glusterfs/2021.03.02/xlator/mount/fuse.so(+0x1de67)[0x7f87715e2e67]
/opt/lib/glusterfs/2021.03.02/xlator/mount/fuse.so(+0x2448c)[0x7f87715e948c]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x9609)[0x7f8772931609]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x43)[0x7f8772856293]
---------
[2021-03-07 19:13:41.311717 +0000] I [MSGID: 100030] [glusterfsd.c:2697:main] 0-/opt/sbin/glusterfs: Started running version [{arg=/opt/sbin/glusterfs}, {version=2021.03.02}, {cmdlinestr=/opt/sbin/glusterfs --process-name fuse -l /var/log/gluster/gluster.log --volfile-id storage-pool-1 --fs-display-name kadalu:storage-pool-1 -f /kadalu/volfiles/storage-pool-1.client.vol /mnt/storage-pool-1}] 
[2021-03-07 19:13:41.311952 +0000] E [fuse-bridge.c:6756:init] 0-fuse: Mountpoint /mnt/storage-pool-1 seems to have a stale mount, run 'umount /mnt/storage-pool-1' and try again.
[2021-03-07 19:13:41.311966 +0000] E [MSGID: 101019] [xlator.c:640:xlator_init] 0-fuse: Initialization of volume failed. review your volfile again. [{name=fuse}] 
[2021-03-07 19:13:42.405592 +0000] I [MSGID: 100030] [glusterfsd.c:2697:main] 0-/opt/sbin/glusterfs: Started running version [{arg=/opt/sbin/glusterfs}, {version=2021.03.02}, {cmdlinestr=/opt/sbin/glusterfs --process-name fuse -l /var/log/gluster/gluster.log --volfile-id storage-pool-1 --fs-display-name kadalu:storage-pool-1 -f /kadalu/volfiles/storage-pool-1.client.vol /mnt/storage-pool-1}] 
[2021-03-07 19:13:42.405812 +0000] E [fuse-bridge.c:6756:init] 0-fuse: Mountpoint /mnt/storage-pool-1 seems to have a stale mount, run 'umount /mnt/storage-pool-1' and try again.
[2021-03-07 19:13:42.405826 +0000] E [MSGID: 101019] [xlator.c:640:xlator_init] 0-fuse: Initialization of volume failed. review your volfile again. [{name=fuse}] 
```

**From the nodeplugin-container:** (repeats indefinitely) 
```
[2021-03-07 19:19:09,806] ERROR [volumeutils - 862:mount_glusterfs] - error to execute command   volume={'name': 'storage-pool-1', 'g_volname': None, 'g_host': None, 'g_options': None, 'type': 'Replica3'} cmd=['/opt/sbin/glusterfs', '--process-name', 'fuse', '-l', '/var/log/gluster/gluster.log', '--volfile-id', 'storage-pool-1', '--fs-display-name', 'kadalu:storage-pool-1', '-f', '/kadalu/volfiles/storage-pool-1.client.vol', '/mnt/storage-pool-1'] error=[1]  
[2021-03-07 19:19:09,806] ERROR [_server - 445:_call_behavior] - Exception calling application: [1]  
Traceback (most recent call last):
  File "/kadalu/lib/python3.8/site-packages/grpc/_server.py", line 435, in _call_behavior
    response_or_iterator = behavior(argument, context)
  File "/kadalu/nodeserver.py", line 72, in NodePublishVolume
    mount_glusterfs(volume, mntdir, True)
  File "/kadalu/volumeutils.py", line 868, in mount_glusterfs
    raise err
  File "/kadalu/volumeutils.py", line 860, in mount_glusterfs
    execute(*cmd)
  File "/kadalu/kadalulib.py", line 121, in execute
    raise CommandException(proc.returncode, out.strip(), err.strip())
kadalulib.CommandException: [1] 
```

**From the node-driver-registrar**:
```
I0306 18:09:31.478508       1 main.go:112] Version: v2.0.1-0-g1aaa13a-dirty
I0306 18:09:31.478547       1 main.go:122] Attempting to open a gRPC connection with: "/plugin/csi.sock"
I0306 18:09:31.478566       1 connection.go:151] Connecting to unix:///plugin/csi.sock
I0306 18:09:32.480650       1 main.go:129] Calling CSI driver to discover driver name
I0306 18:09:32.480677       1 connection.go:180] GRPC call: /csi.v1.Identity/GetPluginInfo
I0306 18:09:32.480683       1 connection.go:181] GRPC request: {}
I0306 18:09:32.484318       1 connection.go:183] GRPC response: {"name":"kadalu","vendor_version":"0.7.0"}
I0306 18:09:32.484375       1 connection.go:184] GRPC error: <nil>
I0306 18:09:32.484383       1 main.go:139] CSI driver name: "kadalu"
I0306 18:09:32.484407       1 node_register.go:55] Starting Registration Server at: /registration/kadalu-reg.sock
I0306 18:09:32.484553       1 node_register.go:64] Registration Server started at: /registration/kadalu-reg.sock
I0306 18:09:32.484691       1 node_register.go:86] Skipping healthz server because port set to: 0
I0306 18:09:33.477210       1 main.go:79] Received GetInfo call: &InfoRequest{}
I0306 18:09:33.496498       1 main.go:89] Received NotifyRegistrationStatus call: &RegistrationStatus{PluginRegistered:true,Error:,}
```

**Comment by amarts** [comment]
> signal received: 11

```
/opt/lib/libglusterfs.so.0(__gf_free+0xb0)[0x7f87729fb440]
/opt/lib/glusterfs/2021.03.02/xlator/cluster/replicate.so(+0x67f40)[0x7f876e4edf40]
/opt/lib/glusterfs/2021.03.02/xlator/cluster/replicate.so(+0x67fc8)[0x7f876e4edfc8]
/opt/lib/glusterfs/2021.03.02/xlator/cluster/replicate.so(+0x67fdd)[0x7f876e4edfdd]
/opt/lib/libglusterfs.so.0(fd_unref+0x11f)[0x7f87729f8c1f]
/opt/lib/glusterfs/2021.03.02/xlator/mount/fuse.so(+0x203d3)[0x7f87715e53d3]
```
I have raised the issue in glusterfs repo, and will check why this is happening. Meantime, as this is not frequently seen, and with your description, would happen once in an hour, can we try to identify the particular activity which would happen ? Say a cronjob or a routing backup or DB backup type of operation? That would allow us to corner it faster!

**Comment by wojtek-viirtue** [comment]
@amarts I can't pinpoint it on one thing specifically since we have around 10 PVCs that are performing read/write operations.  What I can tell you...is that all pods using a PVC have endpoint transport errors and need to be restarted in order to reestablish connectivity to the mounts.

Furthermore, as described earlier...any pods using a PVC hang in a KILLING status most likely because they're trying to unmount from the volumes, but can't because of the endpoint transport issues.

**Comment by amarts** [comment]
Planning to give you an updated gluster version to run kadalu! Once I have the new tags built, will let you know (ETA: 2 days) Meantime, if you get a chance to test, or experiment, see if the same behavior is seen with 'Replica1' setups.

**Comment by wojtek-viirtue** [comment]
Thank you....I will investigate that a bit further.  For the time being, I've spun up cephfs with rook and it's working without issue.

In doing some research, there was some interest in a glusterfs integration with rook.  Looks like there might be some overlap and synergy with kadalu.  Have you guys thought of potentially adapting the operator to work with rook?  https://github.com/rook/rook/issues/2044

Rook seems to be a bit heavy, which is what I love about kadalu and how lightweight it is...but thought I'd offer the suggestion.  Kadalu was working flawlessly until upgrading to the latest version.

**Comment by amarts** [comment]
> Kadalu was working flawlessly until upgrading to the latest version.

That is a good feedback! We had some fear that would happen, as we rebased glusterfs version for making storage pool config simpler with distribute etc. It would allow us to corner the issue sooner.

We wanted to keep minimum layers with kadalu, and hence if you notice, we removed many things which heketi had, and made them all simpler. We removed many glusterfs translators which are not being used too.

IMO, at least for now, we would prefer to focus on getting a good release of glusterfs which works flawlessly for container usecases, and then keep CSI/operator code to be simple in kadalu. But if someone who is well versed with rook can offer 'integration', we are happy to provide guidance.
